packages/shared/lib/sdk/sync.ts

```diff
@@ -743,6 +767,7 @@ export class NangoSync extends NangoAction {
     }
 
     public override async getMetadata<T = Metadata>(): Promise<T> {
+        this.verifySyncState();
```

Comment:
isn't there some sort of decorator in Typescript which could handle adding this call dynamically on all class methods. maybe @bodinsamuel knows

Reply:
Yeah I was going to add a comment on that, unfortunately there is no magic keyword like there on some others languages. 
Like my PR on the tracing, there are alternatives:
- Using proxy and `get()` 
- Create a Child class, e.g: CancelableAction and CancelableSync but it will still need to be handled manually for each methods

Reply:
ChatGPT says using a higher order function

```
function withExitSyncIfAborted<T extends { new (...args: any[]): {} }>(constructor: T) {
    return class extends constructor {
        constructor(...args: any[]) {
            super(...args);
            Object.getOwnPropertyNames(constructor.prototype).forEach(method => {
                if (typeof this[method] === 'function' && method !== 'constructor') {
                    const originalMethod = this[method];
                    this[method] = async (...args: any[]) => {
                        // Assuming `this` has a method `exitSyncItAborted` to check for the aborted condition
                        this.exitSyncItAborted();

                        // Proceed with the original method if not aborted
                        return originalMethod.apply(this, args);
                    };
                }
            });
        }
    };
}
```

but I find that this might be misleading or hard to reason about if it isn't explicit.

On the other hand adding this method check might be forgotten to be added by future developers

Reply:
> On the other hand adding this method check might be forgotten to be added by future developers

aka: me next week :)

Reply:
yeah with the proxy we would cover everything by default, but putting the code elsewhere has its own issues
```ts
new Proxy(rawNango, {
    get(target, propKey) {
          return () => { 
             this.checkSomething();
             return target[propKey]
          }
   }
}
```